File: data_structures/heap.java
// A heap is a special tree-based data structure that satisfies the heap property.
// In a max heap, for any given node, the value of the node is greater than or equal to the values of its children, and the highest value is at the root.
// Conversely, in a min heap, the value of the node is less than or equal to the values of its children, with the smallest value at the root.

// imagine you are organizing a series of tasks based on priority, where the most urgent task needs to be addressed first.
// A max heap could be used to model this sscenario, where the highest-priority task is always at the top of the heap.
// When you complete the most urgent task, you remove it from the heap, and the next most urgent task moves to the top.

// Use Cases
// 1. Priority Queue: Heaps are commonly used to implement priority queues.
// 2. Scheduling Algorithms: In operating systems, heaps can be used to schedule processes.
// 3. Graph Algorithms: Heaps are used in algorithms like Dijkstra's shortest path algorithm to efficiently find the minimum distance.

// A heap is typically implemented using an array, where:
// - The root element is at index 0.
// - For any node at index `i`, the left child is at index `2i + 1` and the right child is at index `2i + 2`
// - The parent of a node at index `i` is at index `(i - 1) / 2`

public class MaxHeap {
    private int[] heap;
    private int size;
    private int maxSize;

    // Constructor to intiialize an empty heap
    public MaxHeap(int maxSize) {
        this.maxSize = maxSize;
        this.size = 0;
        this.heap = new int[this.maxSize];
    }

    // Return the index of the parent of the node at index i
    private int parent(int i) {
        return (it - 1) / 2;
    }

    // Return the index of the left child of the node at index i
    private int leftChild(int i) {
        return (2 * i) + 1;
    }

    // Return the index of the right child of the node at index i
    private int rightChild(int i) {
        return (2 * i) + 2;
    }

    // Insert a new element into the heap
    public void insert(int element) {
        if (size >= maxSize) {
            return;
        }
        heap[size] = element;
        int current = size;

        while (heap[current] > heap[parent(current)]) {
            swap(current, parent(current));
            current = parent(current);
        }
        size++;
    }

    // Swap two element in the heap
    private void swap(int i, int j) {
        int temp = heap[i];
        heap[i] = heap[j];
        heap[j] = temp;
    }

    // Remove and return the maximum element from the heap
    public int extractMax() {
        int popped = heap[0];
        heap[0] = heap[--size];
        maxHeapify(0);
        return popped;
    }

    // Maintain the max heap property
    private void maxHeapify(int j) {
        int left = leftChild(i);
        int right = rightChild(i);
        int largest = i;

        if (left < size && heap[left] > heap[largest]) {
            largest = left;
        }

        if (right < size && heap[right] > heap(largest)) {
            largest = right;
        }

        if (largest != i) {
            swap(i, largest);
            maxHeapify(largest);
        }
    }

    // Print the heap
    public void printHeap() {}
}

File: data_structures/heap.kt
// A heap is a specialized tree-based data structure that satisfies the heap property.

// There are two types of heap:
// 1. Max-Heap: For any given node I, the value of I is greater than or equal to the values of its children.
// 2. Min-Heap: For any given node i, the value of I is less than or equal to the values of its children.

// Think of a priority queue system in a hospital. Patients are assigned priority numbers, and the one with the highest priority (like critical condition) is treated first.
// This system can be implemented using a max-heap where the patient with the maximum priority is always at the top.

class MaxHeap(private val size: Int) {
    private val items = IntArray(size)
    private var count = 0

    // Get index of the left child, right child, and parent of the current node
    private fun getLeftChildIndex(parentIndex: Int) = 2 * parentIndex + 1
    private fun getRightChildIndex(parentIndex: Int) = 2 * parentIndex + 2
    private fun getParentIndex(childIndex: Int) = (childIndex - 1) / 2

    // Check if the current node has a left child, right child, or parent
    private fun hasLeftChild(index: Int) = getLeftChildIndex(index) < count
    private fun hasRightChild(index: Int) = getRightChildIndex(index) < count
    private fun hasParent(index: Int) = getParentIndex(index) >= 0

    // Retrieve the left child, right child, or parent of the current node
    private fun leftChild(index: Int) = items[getLeftChildIndex(index)]
    private fun rightChild(index: Int) = items[getRightChildIndex(index)]
    private fun parent(index: Int) = items[getParentIndex(index)]

    // Get the maximum element without removing it
    fun peek(): Int {
        if (count == 0) throw IllegalStateException()
        return items[0]
    }

    // Remove and return the maximum element
    fun pool(): Int {
        if (count == 0) throw IllegalStateException()
        val item = items[0]
        items[0] = items[count - 1]
        count--
        heapifyDown()
        return item
    }

    // Ad an element to the heap
    fun add(item: Int) {
        items[count] = item
        count++
        heapifyUp()
    }

    // Move the newly added element up to its correct position to maintain the heap property
    private fun heapifyUp() {
        var index = count - 1
        while (hasParent(index) && parent(index) < items[index]) {
            swap(getParentIndex(index), index)
            index = getParentIndex(index)
        }
    }

    // Move the root element down to its correct position after removal of the maximum element
    private fun heapifyDown() {
        var index = 0
        while (hasLeftChild(index)) {
            var largerChildIndex = getLeftChildIndex(index)
            if (hasRightChild(index) && rightChild(index) > leftChild(index)) {
                largerChildIndex = getRightChildIndex(index)
            }
            if (items[index] > items[largerChildIndex]) {
                break
            } else {
                swap(index, largerChildIndex)
            }
            index = largerChildIndex
        }
    }

    // Utility function to swap two elements in the array
    private fun swap(indexOne: Int, indexTwo: Int) {
        val temp = items[indexOne]
        items[indexOne] = items[indexTwo]
        items[indexTwo] = temp
    }
}

fun main() {
    val maxHeap = MaxHeap(10)
    maxHeap.add(5)
    maxHeap.add(3)
    maxHeap.add(8)
    println(maxHeap.peek()) // Output: 8
    println(maxHeap.poll()) // Output: 8
    println(maxHeap.peek()) // Output: 5
}